Firmware/communication/CircularBuffer.hpp

```diff
@@ -0,0 +1,333 @@
+/*******************************************************************************
+* File          : circularBuffer.hpp
+*
+* Description   : Implementation file for circular buffers
+*
+* Project       :
+*
+* Author        : S.Gilbert
+*
+* Created on    : 09 Dec 2019
+*
+*
+*******************************************************************************/
+/* Define to prevent recursive inclusion -------------------------------------*/
+#ifndef CIRCULAR_BUFFER_H
+#define CIRCULAR_BUFFER_H
+
+/*******************************************************************************
+INCLUDES
+*******************************************************************************/
+
+#include <string.h>
+#include <stddef.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+/*******************************************************************************
+NAMESPACE
+*******************************************************************************/
+
+/*******************************************************************************
+DEFINITIONS
+*******************************************************************************/
+
+typedef enum
+{
+	CB_OK = 0,
+	CB_FAIL,
+	CB_SIZE_ERROR,
+	CB_POINTER_ERROR
+}CB_Status_t;
+
+/*******************************************************************************
+TYPES
+*******************************************************************************/
+
+/*******************************************************************************
+GLOBAL VARIABLES
+*******************************************************************************/
+
+/*******************************************************************************
+FUNCTION PROTOTYPES
+*******************************************************************************/
+
+template <class dataType_t>
+class CCBBuffer
+{
+    typedef struct
+    {
+        size_t numElements;                                                     /* maximum number of elements */
+        int32_t start;                                                          /* index of oldest element */
+        int32_t end;                                                            /* index at which to write new element */
+    }CBTracker_t;
+
+    typedef struct
+    {
+        CBTracker_t tracker;                                                    /* buffer size and start/end counters */
+        dataType_t * pArray;                                                    /* pointer to array of elements. to be set on buffer init */
+    }circularBuffer_t;
```

Comment:
What is the reason for nesting these fields inside structs instead of making them top level member fields?

Reply:
No real reason here but was a hold over from when it was C code and not templated. I will flatten the structure here.